### PR TITLE
Singa 255 Compile mandatory depedent libaries together with SINGA code

### PR DESCRIPTION
**Pull request by xiezl**

### CODE REVIEW

**Comment by xiezl** [comment]
If it is configured with option "BUILD_SHARED_LIBS=OFF", singa will find static libs when possible.
I moved all submodules/dependent libs into external cmake projects. Hence, lib folder is gone.

File: CMakeLists.txt [hunk 1]
```diff
@@ -49,11 +50,13 @@ OPTION(USE_CUDA "Use Cuda libs" OFF)
...
 OPTION(USE_OPENCV "Use opencv" OFF)
 OPTION(USE_LMDB "Use LMDB libs" OFF)
-OPTION(USE_PYTHON "Generate py wrappers" ON)
+OPTION(USE_PYTHON "Generate py wrappers" OFF)
 OPTION(USE_JAVA "Generate java wrappers" OFF)
 OPTION(USE_OPENCL "Use OpenCL" OFF)
 OPTION(ENABLE_DIST "enable distributed training" OFF)
 OPTION(DISABLE_WARNINGS "disable warnings under windows" ON)
+OPTION(USE_MODULES "Use submodules to compile" OFF)
+OPTION(BUILD_SHARED_LIBS "Use shared library" OFF)
```
Comment (xiezl) [inline CMakeLists.txt]: use shared libs or build shared libs?
use shared libs: the dependent libs are shared libs
build shared libs: build libsinga.so

File: CMakeLists.txt [hunk 2]
```diff
@@ -71,14 +74,73 @@ SET(LIBRARY_OUTPUT_PATH ${PROJECT_BINARY_DIR}/lib)
 SET(EXECUTABLE_OUTPUT_PATH ${PROJECT_BINARY_DIR}/bin)
 
 IF (USE_CUDA)
-    ADD_SUBDIRECTORY(lib/cnmem)
-    LIST(APPEND SINGA_LINKER_LIBS cnmem)
-    SET(global_cuda_objs "")
+    include(ExternalProject)
+    ExternalProject_Add(cnmem
+        GIT_REPOSITORY "https://github.com/NVIDIA/cnmem.git"
```
Comment (xiezl) [inline CMakeLists.txt]: can we just download from a specific commit?
the master could change later, including the APIs.

File: CMakeLists.txt [hunk 3]
```diff
@@ -71,14 +74,73 @@ SET(LIBRARY_OUTPUT_PATH ${PROJECT_BINARY_DIR}/lib)
...
+    ExternalProject_Add(openblas
+        GIT_REPOSITORY "https://github.com/xianyi/OpenBLAS.git"
+        GIT_TAG "develop"
+        SOURCE_DIR "openblas/" 
+        BUILD_IN_SOURCE 1
+        CONFIGURE_COMMAND ""
+        BUILD_COMMAND "make" "ONLY_CBLAS=1" 
+        INSTALL_COMMAND "make" "PREFIX=${CMAKE_BINARY_DIR}/" "install"
+        )
+    IF(USE_CUDA)
```
Comment (xiezl) [inline CMakeLists.txt]: duplicated? Line 77..

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -95,6 +95,7 @@ ADD_DEPENDENCIES(singa_objects copy_protobuf)
 IF(UNIX OR APPLE)
   ADD_LIBRARY(singa SHARED $<TARGET_OBJECTS:singa_objects> ${cuda_objs})
   TARGET_LINK_LIBRARIES(singa ${SINGA_LINKER_LIBS})
```
Comment (xiezl) [inline src/CMakeLists.txt]: no need to add the dependent libs?

File: cmake/Dependencies.cmake [hunk 5]
```diff
@@ -47,30 +84,23 @@ ELSE()
     SET(USE_CUDNN FALSE)
 ENDIF()
 
-IF(USE_CBLAS)
```
Comment (xiezl) [inline cmake/Dependencies.cmake]: Yes. USE_CBLAS would always be ON. If we remove it, then we need to remove other places using USE_CBLAS

File: cmake/Dependencies.cmake [hunk 6]
```diff
@@ -18,59 +18,94 @@
...
     FIND_PACKAGE(LMDB REQUIRED)
     INCLUDE_DIRECTORIES(SYSTEM ${LMDB_INCLUDE_DIR})
     LIST(APPEND SINGA_LINKER_LIBS ${LMDB_LIBRARIES})
-    MESSAGE(STATUS "FOUND lmdb at ${LMDB_INCLUDE_DIR}")
+    #MESSAGE(STATUS "FOUND lmdb at ${LMDB_INCLUDE_DIR}")
 ENDIF()
 
 IF(USE_CUDA)
     INCLUDE("cmake/Cuda.cmake")
+    SET(CNMEM_INCLUDE_DIR "${CMAKE_BINARY_DIR}/include")
```
Comment (nudles) [inline cmake/Dependencies.cmake]: could you please also compile cnmem into .a?

File: CMakeLists.txt [hunk 7]
```diff
@@ -16,44 +16,49 @@
...
 OPTION(USE_CUDNN "Use Cudnn libs" ON)
 OPTION(USE_OPENCV "Use opencv" OFF)
 OPTION(USE_LMDB "Use LMDB libs" OFF)
-OPTION(USE_PYTHON "Generate py wrappers" ON)
+OPTION(USE_PYTHON "Generate py wrappers" OFF)
 OPTION(USE_JAVA "Generate java wrappers" OFF)
 OPTION(USE_OPENCL "Use OpenCL" OFF)
 OPTION(ENABLE_DIST "enable distributed training" OFF)
 OPTION(DISABLE_WARNINGS "disable warnings under windows" ON)
+OPTION(USE_MODULES "Use submodules to compile" OFF)
```
Comment (nudles) [inline CMakeLists.txt]: "compile dependent libs as submodules together with singa"

**Pull request merged**